base/compiler/sort.jl

```diff
@@ -0,0 +1,99 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+# reference on sorted binary search:
+#   http://www.tbray.org/ongoing/When/200x/2003/03/22/Binary
+
+# index of the first value of vector a that is greater than or equal to x;
+# returns lastindex(v)+1 if x is greater than all values in v.
+function searchsortedfirst(v::AbstractVector, x, lo::T, hi::T, o::Ordering)::keytype(v) where T<:Integer
+    hi = hi + T(1)
+    len = hi - lo
+    @inbounds while len != 0
+        half_len = len >>> 0x01
+        m = lo + half_len
+        if lt(o, v[m], x)
+            lo = m + 1
+            len -= half_len + 1
+        else
+            hi = m
+            len = half_len
+        end
+    end
+    return lo
+end
+
+# index of the last value of vector a that is less than or equal to x;
+# returns firstindex(v)-1 if x is less than all values of v.
+function searchsortedlast(v::AbstractVector, x, lo::T, hi::T, o::Ordering)::keytype(v) where T<:Integer
+    u = T(1)
+    lo = lo - u
+    hi = hi + u
+    @inbounds while lo < hi - u
+        m = midpoint(lo, hi)
+        if lt(o, x, v[m])
+            hi = m
+        else
+            lo = m
+        end
+    end
+    return lo
+end
+
+# returns the range of indices of v equal to x
+# if v does not contain x, returns a 0-length range
+# indicating the insertion point of x
+function searchsorted(v::AbstractVector, x, ilo::T, ihi::T, o::Ordering)::UnitRange{keytype(v)} where T<:Integer
+    u = T(1)
+    lo = ilo - u
+    hi = ihi + u
+    @inbounds while lo < hi - u
+        m = midpoint(lo, hi)
+        if lt(o, v[m], x)
+            lo = m
+        elseif lt(o, x, v[m])
+            hi = m
+        else
+            a = searchsortedfirst(v, x, max(lo,ilo), m, o)
+            b = searchsortedlast(v, x, m, min(hi,ihi), o)
+            return a : b
+        end
+    end
+    return (lo + 1) : (hi - 1)
+end
+
+for s in [:searchsortedfirst, :searchsortedlast, :searchsorted]
+    @eval begin
+        $s(v::AbstractVector, x, o::Ordering) = $s(v,x,firstindex(v),lastindex(v),o)
+        $s(v::AbstractVector, x;
+           lt=isless, by=identity, rev::Union{Bool,Nothing}=nothing, order::Ordering=Forward) =
+            $s(v,x,ord(lt,by,rev,order))
+    end
+end
+
+# An unstable sorting algorithm for constructing IncrementalCompact.
+function sort!(v::Vector; by::Function=identity, (<)::Function=<)
+    isempty(v) && return v # This branch is hit 95% of the time
+
+    if length(v) > 30 # Comb pass avoids quadratic runtime
+        interval = (3 * length(v)) >> 2
+        while interval > 1
+            @inbounds for j in 1:length(v)-interval
+                a, b = v[j], v[j+interval]
+                v[j], v[j+interval] = by(b) < by(a) ? (b, a) : (a, b)
+            end
+            interval = (3 * interval) >> 2
+        end
+    end
```

Comment:
The distribution of sorting tasks generated from building julia and loading Makie is skewed heavily toward empty and short vectors. The data from https://github.com/JuliaLang/julia/pull/45584 is suggestive, though slightly out of date. Because of this, Insertion sort seems like it should definitely be a component.

I want to avoid slipping an accidental quadratic runtime into the compilation of large or unusual functions, so I added a comb pass which makes the algorithm empirically close to n log n runtime for random input:

```julia
julia> for n in 1:7
           t = @belapsed Core.Compiler.sort!(x; <) setup=(x=rand($(10^n)))
           println(n, ": ", t/10^n*1e9)
       end
1: 1.3267803410230692
2: 7.918383838383839
3: 14.363
4: 20.0192
5: 29.841300000000004
6: 54.020132999999994
7: 69.97564779999999
```

Theoretical worst cases are not so bad as they are vanishingly unlikely and practically unreached. For example, `sort!(rand(n); by=identity)` currently has quadratic worst case runtime but nobody has opened an issue about it to my knowledge.

Nevertheless, the cost of switching to heapsort is pretty small (a bit of complexity and about .2ms extra runtime building julia and loading Makie).

Do folks think this change is worth making?

```suggestion
    # Of the remaining 5%, this branch is hit less than 1% of the time
    if length(v) > 200 # Heap sort prevents quadratic runtime
        o = ord(<, by, true)
        heapify!(v, o)
        for i in lastindex(v):-1:2
            y = v[i]
            v[i] = v[1]
            percolate_down!(v, 1, y, o, i-1)
         end
        return v
    end
```

It's still unstable and still doesn't allocate. I'm mildly in favor.

Reply:
I'd say the heap sort is probably the better option on the grounds that I'd feel really stupid if we were tracking down a huge compiler time in the future and found out it was because of the sorting algorithm we were using.